nixos/modules/services/development/zammad.nix

```diff
@@ -0,0 +1,217 @@
+      };
+
+      websocketPort = lib.mkOption {
+        type = lib.types.int;
+        default = 6042;
+        description = "Websocket service port.";
+      };
+
+      dbName = lib.mkOption {
+        type = lib.types.str;
+        default = "zammad";
+        description = "The name of the database to use.";
+      };
+
+      dbUsername = lib.mkOption {
+        type = lib.types.str;
+        default = "zammad";
+        description = "The username to use to connect to the database.";
+      };
+
+      secretsFile = lib.mkOption {
+        type = lib.types.nullOr lib.types.path;
+        default = null;
+        description = ''
+          Path of a file containing secrets the format of EnvironmentFile as
+          described by systemd.exec(5). You must to define:
+            - PGPASSWORD
+            - SECRET_KEY_BASE
+          SECRET_KEY_BASE can be generated using:
+            ruby -e "require 'securerandom'; puts SecureRandom.hex(64)"
+        '';
+      };
+    };
+
+  };
+
+  config = lib.mkIf cfg.enable {
+
+    networking.firewall.allowedTCPPorts = [
+      config.services.zammad.port
+      config.services.zammad.websocketPort
+    ];
+
+    users.users.zammad = {
+      isSystemUser = true;
+      home = cfg.dataDir;
+      group = "zammad";
+    };
+
+    users.groups.zammad = {};
+
+    services.postgresql = {
+      enable = true;
+      ensureDatabases = [ cfg.dbName ];
+      ensureUsers = [
+        {
+          name = cfg.dbUsername;
+          ensurePermissions."DATABASE ${cfg.dbName}" = "ALL PRIVILEGES";
+        }
+      ];
+    };
+
+    systemd.services.zammad-setup = {
+      after = [ "postgresql.service" ];
+      requires = [ "postgresql.service" ];
+      serviceConfig.EnvironmentFile = cfg.secretsFile;
+      script = ''
+        echo "Setting password for database '${cfg.dbName}' and user '${cfg.dbUsername}'"
+        ${pkgs.utillinux}/bin/runuser -u ${config.services.postgresql.superUser} -- \
+          ${config.services.postgresql.package}/bin/psql \
+            -c "ALTER ROLE ${cfg.dbUsername} WITH PASSWORD '$PGPASSWORD'"
+        mkdir -p ${cfg.dataDir}
+      '';
+    };
+
+    systemd.services.zammad-web = {
+      after = [
+        "network.target"
+        "postgresql.service"
+        "zammad-setup.service"
+      ];
+      requires = [
+        "postgresql.service"
+        "zammad-setup.service"
+      ];
+      description = "Zammad web";
+      wantedBy = [ "multi-user.target" ];
+      environment = env;
+      preStart = ''
+        # Blindly copy the whole project here.
+        chmod -R +w ${cfg.dataDir}/
+        rm -rf ${cfg.dataDir}/public/assets/*
+        rm -rf ${cfg.dataDir}/tmp/*
+        rm -rf ${cfg.dataDir}/log/*
+        cp -r --no-preserve=owner ${cfg.package}/* ${cfg.dataDir}
+        chmod -R +w ${cfg.dataDir}
+        export DATABASE_URL="postgresql://${cfg.dbUsername}:$PGPASSWORD@localhost:${toString(config.services.postgresql.port)}/${cfg.dbName}";
```

Comment:
since this is used multiple times it might make sense to move it out. also another instance of special character problems in PGPASSWORD.